File: 494-target-sum/target-sum.cpp
class Solution {
public:
    void bc(int eL, vector<int>& nums, int target, int& sum){
        if(eL == 0 && target == 0){
            sum++;
            return;
        }

        if(eL>0) {bc(eL-1, nums, target - nums[eL-1], sum);
        bc(eL-1, nums, target + nums[eL-1], sum);}
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        int sum = 0;
        bc(nums.size(), nums, target, sum);
        return sum;
    }
};

File: 494-target-sum/target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        ln = len(nums)
        dp = {}

        def recurs(i, target):
            nonlocal dp, ln, nums
            if i>ln:
                return 0
            if i == ln:
                return 1 if target==0 else 0
            if (i, target) in dp:
                return dp[(i, target)]
            
            dp[(i, target)] = recurs(i+1, target+nums[i]) + recurs(i+1, target-nums[i])
            return dp[(i, target)]

        return recurs(0, target)
        